### PR TITLE
patch for NVTX

### PR DESCRIPTION
**Pull request by anaruse**
I've added some edit for NVIDIA Tools Extension Library (NVTX).  NVTX makes it easy to do some performance analysis or optimization work using nvprof or nvvp.  It is mainly for GPU but useful for CPU as well.  Please check it when you have a time.
BR.

### CODE REVIEW

**Comment by anaruse** [comment]
added some more change to support color choice.

File: cupy/cuda/nvtx.pyx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+colors[3] = 0x0000FF
+colors[4] = 0x7F00FF
+colors[5] = 0xFF007F
+colors[6] = 0xFF0000
+colors[7] = 0xFF7F00
+colors[8] = 0x7FFF00
+colors[9] = 0x7F7F7F
+
+
+cpdef void Mark(str message, int id_color=-1) except *:
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: How about splitting the function?
One function receive a color code, other function receive a color id.

File: cupy/cuda/nvtx.pyx [hunk 2]
```diff
@@ -0,0 +1,84 @@
+"""
+Wrapper for NVIDIA Tools Extension Library (NVTX)
+
+"""
+cimport cython
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: Please remove it.
This import is not used.

File: cupy/cuda/nvtx.pyx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    attrib[0].colorType = NVTX_COLOR_ARGB
+    attrib[0].color = colors[id_color % num_colors]
+
+    nvtxMarkEx(attrib)
+
+
+cpdef void RangePush(str message, int id_color=-1) except *:
+    cdef bytes b_message = message.encode()
+    if id_color < 0 or NVTX_VERSION != 1:
+        ret = nvtxRangePushA(<const char*>b_message)
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: Do you use `ret`?

File: cupy/cuda/nvtx.pyx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        attrib[0].message.ascii = NULL
+
+    attrib[0].version = NVTX_VERSION
+    attrib[0].size = NVTX_EVENT_ATTRIB_STRUCT_SIZE
+    attrib[0].messageType = NVTX_MESSAGE_TYPE_ASCII
+    attrib[0].message.ascii = b_message
+    attrib[0].colorType = NVTX_COLOR_ARGB
+    attrib[0].color = colors[id_color % num_colors]
+
+    ret = nvtxRangePushEx(attrib)
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: Do you use `ret`?

File: cupy/cuda/nvtx.pyx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    # assert ret >= 0, 'nvtxRangePushEx(): {}'.format(ret)
+
+
+cpdef void RangePop() except *:
+    ret = nvtxRangePop()
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: Do you use `ret`?

File: cupy/cuda/nvtx.pyx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+colors[9] = 0x7F7F7F
+
+
+cpdef void Mark(str message, int id_color=-1) except *:
+    cdef bytes b_message = message.encode()
+    if id_color < 0 or NVTX_VERSION != 1:
+        nvtxMarkA(<const char*>b_message)
+        return
+
+    cdef nvtxEventAttributes_t attrib[1]
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: How about following code?
```python
    cdef nvtxEventAttributes_t attrib
    if NVTX_VERSION == 1:
        attrib.version = 0
        attrib.size = 0
        attrib.category = 0
        attrib.colorType = 0
        attrib.color = 0
        attrib.payloadType = 0
        attrib.reserved0 = 0
        attrib.payload.ullValue = 0
        attrib.messageType = 0
        attrib.message.ascii = NULL

    attrib.version = NVTX_VERSION
    attrib.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE
    attrib.messageType = NVTX_MESSAGE_TYPE_ASCII
    attrib.message.ascii = b_message
    attrib.colorType = NVTX_COLOR_ARGB
    attrib.color = colors[id_color % num_colors]

    nvtxMarkEx(&attrib)
```

File: cupy/cuda/nvtx.pyx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    if NVTX_VERSION == 1:
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: This statement is always true.

File: cupy/cuda/nvtx.pyx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        attrib[0].version = 0
+        attrib[0].size = 0
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: How about using ``memset``?
```python
from libc cimport string
cdef nvtxEventAttributes_t attrib
string.memset(&attrib, 0, sizeof(attrib))
```

File: cupy/cuda/nvtx.pyx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+
+cpdef void RangePush(str message, int id_color=-1) except *:
+    cdef bytes b_message = message.encode()
+    if id_color < 0 or NVTX_VERSION != 1:
+        ret = nvtxRangePushA(<const char*>b_message)
+        # assert ret >= 0, 'nvtxRangePushA(): {}'.format(ret)
+        return
+
+    cdef nvtxEventAttributes_t attrib[1]
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: ditto

File: cupy/cuda/nvtx.pyx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    if NVTX_VERSION == 1:
```
Comment (okuta) [inline cupy/cuda/nvtx.pyx]: ditto

File: cupy/cuda/nvtx.pxd [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    ctypedef unsigned long long  uint64_t
+    ctypedef          void        wchar_t
+    cdef union payload_t:
+        uint64_t  ullValue
+        int64_t    llValue
+        double      dValue
+    cdef union message_t:
+        char*     ascii
+        wchar_t*  unicode
+    cdef struct nvtxEventAttributes_v1:
```
Comment (okuta) [inline cupy/cuda/nvtx.pxd]: `nvtxEventAttributes_v1` is almost same `nvtxEventAttributes_v2`.
Only `payload_t` is different.
```python
union payload_t
{
    uint64_t ullValue;
    int64_t llValue;
    double dValue;
    /* NVTX_VERSION_2 */
    uint32_t uiValue;
    int32_t iValue;
    float fValue;
} payload;
```

`cupy.cuda,nvtx,Mark` function do not use `payload`.
I think it is possible to use the same struct in version 1 and 2.

**Comment by okuta** [comment]
Thank you for your PR.
Sorry for late reply.

Please add documents and tests.
The NVTX has two version now. We need tests of calling the NVTX functions.

**Comment by anaruse** [comment]
Thank you for your comments, Okuta-san.

I've updated the NVTX patch based on your comments such as removing unused imports and variables, initializing structured variables using memset(), adding documents and test script and etc.  I've also confirmed it runs OK with NVTX version 2 of CUDA 8.0 RC. Please check it when you get a chance.

BR.

File: cupy/cuda/nvtx.pxd [hunk 12]
```diff
@@ -0,0 +1,53 @@
+"""
+Wrapper for NVIDIA Tools Extension Library (NVTX)
+
+"""
+cdef extern from "cupy_cuda.h":
```
Comment (okuta) [inline cupy/cuda/nvtx.pxd]: Please move this section to `nvtx.pyx`
This section is used by only `nvtx.pyx`.

File: tests/cupy_tests/cuda_tests/test_nvtx.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
+from chainer.testing import attr
+from cupy.cuda import nvtx
+import unittest
+
+
+class TestNVTX(unittest.TestCase):
+    @attr.gpu
```
Comment (okuta) [inline tests/cupy_tests/cuda_tests/test_nvtx.py]: Please add empty line.

**Comment by okuta** [comment]
LGTM!

**Comment by okuta** [comment]
Thank you!

**Pull request merged**